clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java

```diff
@@ -77,6 +78,17 @@
 @PowerMockIgnore("javax.management.*")
 public class KafkaProducerTest {
 
+    @Test
+    public void testMetricsReporterAutoGeneratedClientId() {
+        Properties props = new Properties();
+        props.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        props.setProperty(ProducerConfig.METRIC_REPORTER_CLASSES_CONFIG, MockClientIdMetricsReporter.class.getName());
+        KafkaProducer<String, String> producer = new KafkaProducer<>(
+                props, new StringSerializer(), new StringSerializer());
+        Assert.assertTrue(MockClientIdMetricsReporter.clientId.startsWith("producer-"));
```

Comment:
Can you check that the auto-generated client ID appears here?

Reply:
It is **loosely** checking the auto-generated client ID by checking for the prefix `producer-`. This is because there are many other tests that create clients without specifying the `client.id`, which increments the static `CONSUMER_CLIENT_ID_SEQUENCE` and `PRODUCER_CLIENT_ID_SEQUENCE`. This is problematic as the auto-generated client ID value then depends on the order of test execution. 

There are no getters/setters for this client ID sequence so it is not easy to test for exact auto-generated client ID.

I see two options here:

**(1) Make all other tests set a client ID**
We could make all other tests set the `client.id` field so that the auto-generated client ID will be suffixed with `-1` (producer-1, consumer-1). All future tests would have to follow this and set client ids to ensure these tests always pass (if tests are not always executed in the same order, then we could get into a bad situation where it fails sometimes).

**(2) Create getter methods to mock**
We could create getter methods `getConsumerClientIdSequence()` and `getProducerClientIdSequence()` in `KafkaConsumer` and `KafkaProducer`, and mock them in the unit tests. This will allow us to check for the exact value.

Thoughts? @ijuma

Reply:
Can't you check KafkaConsumer#clientId?  We could add a package-private accessor only to be used for testing.

Reply:
There is no clientId getter currently, but it would indeed be easier if we had a package-private accessor for testing. 

Would there be any point in exposing this getter as public?

Reply:
Good question.  I can't think of any right now (maybe someone else can?) so I would lean towards keeping it package-private for now.